cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll

```diff
@@ -907,21 +994,190 @@ class CastNode extends Node {
   CastNode() { none() } // stub implementation
 }
 
+cached
+newtype TDataFlowCallable =
+  TSourceCallable(Cpp::Declaration decl) or
+  /*{ not decl instanceof FlowSummaryImpl::Public::SummarizedCallable }*/ TSummarizedCallable(
+    // TODO: figure this out
+    FlowSummaryImpl::Public::SummarizedCallable c
+  )
```

Comment:
Yeah, there's a totally legit question here to be answered: If we have a model for a function (i.e., it's a summarized callable) that we _also_ have source code for (i.e., it's a source callable). What kind of callable should it be?

Different languages take different approaches and apply different heuristics here. For C/C++ I think we should just stick to a simple initial rule and say that any function that has a model shouldn't also be a source callable.

How does that sound?

Note that this is not totally in line with what we do for our current modelling (i.e., if we have the source code for `std::vector<int>::push_back` we'll do dataflow through the implementation _and_ using our models which is quite wasteful. I've been wanting to fix this for a while, but it's currently "blocked" by some of our models being less powerful than what we actually infer based on looking at the source code 💪

Reply:
> I think we should just stick to a simple initial rule and say that any function that has a model shouldn't also be a source callable.

My gut feeling was that allowing it to be both is preferable, but you make a case that making modelled functions not a source callable might be more performant.  It might also help to reduce path duplication and other complications.  And it's closer to Swift, which is my baseline here.

Trouble is, if I include that `not decl instanceof FlowSummaryImpl::Public::SummarizedCallable` I mysteriously lose almost all of my results in tests!  I think this may be indicative of something wrong elsewhere.

Reply:
> My gut feeling was that allowing it to be both is preferable, but you make a case that making modelled functions not a source callable might be more performant. It might also help to reduce path duplication and other complications. And it's closer to Swift, which is my baseline here.

Indeed, having both would mean maximum potential flow. That is, as long as one of the cases (i.e., the source code or the model) gives the flow, then we'll report a flow.

One problem is path duplication (as you say).

It's also very possible that dataflow on the source code will fail to realize that a piece of data is actually cleared, whereas the MaD model may make that obvious. And in that case, having both cases means we'll get a FP.

> Trouble is, if I include that not decl instanceof FlowSummaryImpl::Public::SummarizedCallable I mysteriously lose almost all of my results in tests! I think this may be indicative of something wrong elsewhere.

Hm. That's a surprise to me. Let me know if this is something you need help investigating.

Reply:
Fixed.  There were a couple of places where we incorrectly depended on something being a source callable, and deduplicating the `TDataFlowCallable`s surfaced those issues.